Cilium: initial NAT46/64 implementation

borkmann opened the pull request:
See commit msgs. This work adds NAT46 and NAT64 for services to the datapath (tc & XDP).

> * How the NAT46x64 services are provisioned. What is the content of lbmaps for each NAT policy.

```
NAT46 example:

  Note the v4-in-v6 main service entry with the dummy v4 service entry.
  The latter directs the datapath to nat and loop into the v6 one.

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.10:9153   ClusterIP
  2    10.96.0.10:53     ClusterIP
  3    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443

  # ./cilium/cilium service update --frontend="1.1.1.1:8080" --backends="[f00d::2]:6443" --id 7 --k8s-external
  Creating new service with id '7'
  Added service with 1 backends

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.10:9153   ClusterIP
  2    10.96.0.10:53     ClusterIP
  3    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443
  7    1.1.1.1:8080      ExternalIPs    1 => [f00d::2]:6443

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS   BACKEND ADDRESS
  10.96.0.1:443     192.168.178.91:6443 (3)
                    0.0.0.0:0 (3) [ClusterIP, non-routable]
  10.96.0.10:9153   0.0.0.0:0 (1) [ClusterIP, non-routable]
  10.96.0.10:53     0.0.0.0:0 (2) [ClusterIP, non-routable]
  1.1.1.1:8080      0.0.0.0:0 (7) [ExternalIPs, 46x64]
  [1.1.1.1]:8080    [f00d::2]:6443 (7)
                    [::]:0 (7) [ExternalIPs, 46x64]

  # ./cilium/cilium bpf lb list --frontends
  SERVICE ADDRESS   BACKEND ID
  10.96.0.1:443     1 0 (3) [0x0 0x0]
                    0 1 (3) [0x0 0x0]
  10.96.0.10:9153   0 0 (1) [0x0 0x0]
  10.96.0.10:53     0 0 (2) [0x0 0x0]
  1.1.1.1:8080      0 0 (7) [0x41 0x2]
  [1.1.1.1]:8080    2 0 (7) [0x0 0x0]
                    0 1 (7) [0x41 0x2]

  # ./cilium/cilium bpf lb list --backends
  ID   BACKEND ADDRESS
  1    ANY://192.168.178.91:6443
  2    ANY://[f00d::2]:6443

  # ./cilium/cilium service delete 7
  Service 7 deleted successfully

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.10:9153   ClusterIP
  2    10.96.0.10:53     ClusterIP
  3    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS   BACKEND ADDRESS
  10.96.0.1:443     192.168.178.91:6443 (3)
                    0.0.0.0:0 (3) [ClusterIP, non-routable]
  10.96.0.10:9153   0.0.0.0:0 (1) [ClusterIP, non-routable]
  10.96.0.10:53     0.0.0.0:0 (2) [ClusterIP, non-routable]

  # ./cilium/cilium bpf lb list --frontends
  SERVICE ADDRESS   BACKEND ID
  10.96.0.1:443     1 0 (3) [0x0 0x0]
                    0 1 (3) [0x0 0x0]
  10.96.0.10:9153   0 0 (1) [0x0 0x0]
  10.96.0.10:53     0 0 (2) [0x0 0x0]

  # ./cilium/cilium bpf lb list --backends
  ID   BACKEND ADDRESS
  1    ANY://192.168.178.91:6443

NAT64 example:

  Note the v4-in-v6 plus v4 backend entry. The former is used for the
  main DNAT. The v4 is also present for the case where multiple services
  can use a given backend, including v4->v4 services.

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443
  2    10.96.0.10:9153   ClusterIP
  3    10.96.0.10:53     ClusterIP

  # ./cilium/cilium service update --frontend="[f00d::1]:8080" --backends="2.2.2.2:6443" --id 8 --k8s-external
  Creating new service with id '8'
  Added service with 1 backends

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443
  2    10.96.0.10:9153   ClusterIP
  3    10.96.0.10:53     ClusterIP
  8    [f00d::1]:8080    ExternalIPs    1 => 2.2.2.2:6443

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS   BACKEND ADDRESS
  10.96.0.10:9153   0.0.0.0:0 (2) [ClusterIP, non-routable]
  10.96.0.1:443     0.0.0.0:0 (1) [ClusterIP, non-routable]
                    192.168.178.91:6443 (1)
  10.96.0.10:53     0.0.0.0:0 (3) [ClusterIP, non-routable]
  [f00d::1]:8080    [2.2.2.2]:6443 (8)
                    [::]:0 (8) [ExternalIPs, 46x64]

  # ./cilium/cilium bpf lb list --frontends
  SERVICE ADDRESS   BACKEND ID
  10.96.0.10:9153   0 0 (2) [0x0 0x0]
  10.96.0.1:443     0 1 (1) [0x0 0x0]
                    1 0 (1) [0x0 0x0]
  10.96.0.10:53     0 0 (3) [0x0 0x0]
  [f00d::1]:8080    4 0 (8) [0x0 0x0]
                    0 1 (8) [0x41 0x2]

  # ./cilium/cilium bpf lb list --backends
  ID   BACKEND ADDRESS
  1    ANY://192.168.178.91:6443
  4    ANY://2.2.2.2:6443
       ANY://[2.2.2.2]:6443

  # ./cilium/cilium service delete 8
  Service 8 deleted successfully

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443
  2    10.96.0.10:53     ClusterIP
  3    10.96.0.10:9153   ClusterIP

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS   BACKEND ADDRESS
  10.96.0.1:443     192.168.178.91:6443 (1)
                    0.0.0.0:0 (1) [ClusterIP, non-routable]
  10.96.0.10:53     0.0.0.0:0 (2) [ClusterIP, non-routable]
  10.96.0.10:9153   0.0.0.0:0 (3) [ClusterIP, non-routable]

  # ./cilium/cilium bpf lb list --frontends
  SERVICE ADDRESS   BACKEND ID
  10.96.0.10:9153   0 0 (3) [0x0 0x0]
  10.96.0.1:443     1 0 (1) [0x0 0x0]
                    0 1 (1) [0x0 0x0]
  10.96.0.10:53     0 0 (2) [0x0 0x0]

  # ./cilium/cilium bpf lb list --backends
  ID   BACKEND ADDRESS
  1    ANY://192.168.178.91:6443
```

> * Packet path for NAT46.

We enter into the main IPv4 receive path for NodePort, do the service lookup there. Then, we discover that the svc is a NAT46 service. We migrate the packet into an IPv6 one (both src/dst as IPv4-in-IPv6), and recircle into the main IPv6 receive path to do the rest of the NodePort DNAT/SNAT combo there (as if it was a regular IPv6-to-IPv6 NodePort request). Once done, we push the IPv6 packet to the backend.

> * Packet path for NAT64.

We enter into the main IPv6 receive path for NodePort, do the service lookup and pick a backend from the IPv6 backend map. The backends are stored as IPv4-in-IPv6. After DNAT, we do the SNAT with the device's IPv4-in-IPv6 address, so that both src and dst are IPv4-in-IPv6 now. Once done, we migrate the packet to an IPv4 packet and push to the backend.

TODOs beyond these (outside of this PR):
- [ ] pmtu discovery for 4->6
- [ ] k8s svc integration w/ pods

borkmann committed:
bpf: Add initial NAT 4->6 datapath implementation

Manual 4->6 example setup with tcpdump extract of traffic being processed
on the LB node:

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443
  2    10.96.0.10:53     ClusterIP
  3    10.96.0.10:9153   ClusterIP

  # ./cilium/cilium service update --frontend="192.168.178.91:8080" --backends="[2a03:2880:f11c:8183:face:b00c:0:25de]:80" --id 8 --k8s-external
  Creating new service with id '8'
  Added service with 1 backends

  # ./cilium/cilium service list
  ID   Frontend              Service Type   Backend
  1    10.96.0.1:443         ClusterIP      1 => 192.168.178.91:6443
  2    10.96.0.10:53         ClusterIP
  3    10.96.0.10:9153       ClusterIP
  8    192.168.178.91:8080   ExternalIPs    1 => [2a03:2880:f11c:8183:face:b00c:0:25de]:80

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS         BACKEND ADDRESS
  10.96.0.10:9153         0.0.0.0:0 (3) [ClusterIP, non-routable]
  10.96.0.1:443           0.0.0.0:0 (1) [ClusterIP, non-routable]
                          192.168.178.91:6443 (1)
  10.96.0.10:53           0.0.0.0:0 (2) [ClusterIP, non-routable]
  192.168.178.91:8080     0.0.0.0:0 (8) [ExternalIPs, 46x64]
  [192.168.178.91]:8080   [::]:0 (8) [ExternalIPs, 46x64]
                          [2a03:2880:f11c:8183:face:b00c:0:25de]:80 (8)

  # tcpdump -i enp5s0 port 80 or port 8080 -n
  [...]
  13:14:43.918327 IP 192.168.178.92.37702 > 192.168.178.91.8080: Flags [S], seq 3599416146, win 42340, options [mss 1460,sackOK,TS val 991279130 ecr 0,nop,wscale 9], length 0
  13:14:43.918354 IP6 2001:1620:665:0:1ac0:4dff:fe09:c164.37702 > 2a03:2880:f11c:8183:face:b00c:0:25de.80: Flags [S], seq 3599416146, win 42340, options [mss 1460,sackOK,TS val 991279130 ecr 0,nop,wscale 9], length 0
  13:14:43.932886 IP6 2a03:2880:f11c:8183:face:b00c:0:25de.80 > 2001:1620:665:0:1ac0:4dff:fe09:c164.37702: Flags [S.], seq 1236878493, ack 3599416147, win 65535, options [mss 1392,sackOK,TS val 2804102865 ecr 991279130,nop,wscale 8], length 0
  13:14:43.932894 IP 192.168.178.91.8080 > 192.168.178.92.37702: Flags [S.], seq 1236878493, ack 3599416147, win 65535, options [mss 1392,sackOK,TS val 2804102865 ecr 991279130,nop,wscale 8], length 0
  [...]

From external client, connecting to LB node which then performs the above IPv6
request to the backend:

  # curl 192.168.178.91:8080 --verbose
  *   Trying 192.168.178.91:8080...
  * TCP_NODELAY set
  * Connected to 192.168.178.91 (192.168.178.91) port 8080 (#0)
  > GET / HTTP/1.1
  > Host: 192.168.178.91:8080
  > User-Agent: curl/7.68.0
  > Accept: */*
  >
  * Mark bundle as not supporting multiuse
  < HTTP/1.1 301 Moved Permanently
  < Location: http://www.facebook.com:8080/
  < Content-Type: text/html; charset="utf-8"
  < X-FB-Debug: YA5YJcV4uwQfLQOfcWCbcnx3+2ScIzY0l4KEkQc/3/FFkjb/d4n6ApEfZHKqPeLil8EyvWeajvO6twX/z+Raww==
  < Date: Thu, 10 Feb 2022 13:14:43 GMT
  < Priority: u=3,i
  < Alt-Svc: h3=":443"; ma=3600, h3-29=":443"; ma=3600
  < Connection: keep-alive
  < Content-Length: 0
  <
  * Connection #0 to host 192.168.178.91 left intact

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: Move v4-in-v6 address generation code into nat_46x64 header

... given this is going to be reused by code outside of the socket LB.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Remove legacy agent-side NAT46 config

Remove the old/legacy NAT46 endpoint related agent setting. It does not
integrate with Kubernetes and was implemented back at a time where we also
had the Docker plugin. There is little point in keeping it going forward as
we plan to properly integrate NAT46/64 handling for services.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: Remove legacy NAT46 datapath handling

Remove the old/legacy NAT46 datapath side. The helper code in bpf/lib/nat46.h
is still left intact as well as the ENABLE_NAT46 config define. Both will be
reused/extended for NAT46/64 service handling.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: Rename ENABLE_NAT46 into ENABLE_NAT_46X64

... and similarly, rename the header file as it will cover to handle
both NAT46 and NAT64 directions.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: Clean up and generalize ipv4_to_ipv6/ipv6_to_ipv4 functions

Early spring clean up on the old NAT46/64 handling code in order to make
it more generic. Allow to pass target source/destination addresses into
the two ipv4_to_ipv6() / ipv6_to_ipv4() helpers.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Initial implementation of agent NAT46/64 handling

NAT46 example:

  Note the v4-in-v6 main service entry with the dummy v4 service entry.
  The latter directs the datapath to nat and loop into the v6 one.

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.10:9153   ClusterIP
  2    10.96.0.10:53     ClusterIP
  3    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443

  # ./cilium/cilium service update --frontend="1.1.1.1:8080" --backends="[f00d::2]:6443" --id 7 --k8s-external
  Creating new service with id '7'
  Added service with 1 backends

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.10:9153   ClusterIP
  2    10.96.0.10:53     ClusterIP
  3    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443
  7    1.1.1.1:8080      ExternalIPs    1 => [f00d::2]:6443

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS   BACKEND ADDRESS
  10.96.0.1:443     192.168.178.91:6443 (3)
                    0.0.0.0:0 (3) [ClusterIP, non-routable]
  10.96.0.10:9153   0.0.0.0:0 (1) [ClusterIP, non-routable]
  10.96.0.10:53     0.0.0.0:0 (2) [ClusterIP, non-routable]
  1.1.1.1:8080      0.0.0.0:0 (7) [ExternalIPs, 46x64]
  [1.1.1.1]:8080    [f00d::2]:6443 (7)
                    [::]:0 (7) [ExternalIPs, 46x64]

  # ./cilium/cilium bpf lb list --frontends
  SERVICE ADDRESS   BACKEND ID
  10.96.0.1:443     1 0 (3) [0x0 0x0]
                    0 1 (3) [0x0 0x0]
  10.96.0.10:9153   0 0 (1) [0x0 0x0]
  10.96.0.10:53     0 0 (2) [0x0 0x0]
  1.1.1.1:8080      0 0 (7) [0x41 0x2]
  [1.1.1.1]:8080    2 0 (7) [0x0 0x0]
                    0 1 (7) [0x41 0x2]

  # ./cilium/cilium bpf lb list --backends
  ID   BACKEND ADDRESS
  1    ANY://192.168.178.91:6443
  2    ANY://[f00d::2]:6443

  # ./cilium/cilium service delete 7
  Service 7 deleted successfully

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.10:9153   ClusterIP
  2    10.96.0.10:53     ClusterIP
  3    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS   BACKEND ADDRESS
  10.96.0.1:443     192.168.178.91:6443 (3)
                    0.0.0.0:0 (3) [ClusterIP, non-routable]
  10.96.0.10:9153   0.0.0.0:0 (1) [ClusterIP, non-routable]
  10.96.0.10:53     0.0.0.0:0 (2) [ClusterIP, non-routable]

  # ./cilium/cilium bpf lb list --frontends
  SERVICE ADDRESS   BACKEND ID
  10.96.0.1:443     1 0 (3) [0x0 0x0]
                    0 1 (3) [0x0 0x0]
  10.96.0.10:9153   0 0 (1) [0x0 0x0]
  10.96.0.10:53     0 0 (2) [0x0 0x0]

  # ./cilium/cilium bpf lb list --backends
  ID   BACKEND ADDRESS
  1    ANY://192.168.178.91:6443

NAT64 example:

  Note the v4-in-v6 plus v4 backend entry. The former is used for the
  main DNAT. The v4 is also present for the case where multiple services
  can use a given backend, including v4->v4 services.

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443
  2    10.96.0.10:9153   ClusterIP
  3    10.96.0.10:53     ClusterIP

  # ./cilium/cilium service update --frontend="[f00d::1]:8080" --backends="2.2.2.2:6443" --id 8 --k8s-external
  Creating new service with id '8'
  Added service with 1 backends

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443
  2    10.96.0.10:9153   ClusterIP
  3    10.96.0.10:53     ClusterIP
  8    [f00d::1]:8080    ExternalIPs    1 => 2.2.2.2:6443

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS   BACKEND ADDRESS
  10.96.0.10:9153   0.0.0.0:0 (2) [ClusterIP, non-routable]
  10.96.0.1:443     0.0.0.0:0 (1) [ClusterIP, non-routable]
                    192.168.178.91:6443 (1)
  10.96.0.10:53     0.0.0.0:0 (3) [ClusterIP, non-routable]
  [f00d::1]:8080    [2.2.2.2]:6443 (8)
                    [::]:0 (8) [ExternalIPs, 46x64]

  # ./cilium/cilium bpf lb list --frontends
  SERVICE ADDRESS   BACKEND ID
  10.96.0.10:9153   0 0 (2) [0x0 0x0]
  10.96.0.1:443     0 1 (1) [0x0 0x0]
                    1 0 (1) [0x0 0x0]
  10.96.0.10:53     0 0 (3) [0x0 0x0]
  [f00d::1]:8080    4 0 (8) [0x0 0x0]
                    0 1 (8) [0x41 0x2]

  # ./cilium/cilium bpf lb list --backends
  ID   BACKEND ADDRESS
  1    ANY://192.168.178.91:6443
  4    ANY://2.2.2.2:6443
       ANY://[2.2.2.2]:6443

  # ./cilium/cilium service delete 8
  Service 8 deleted successfully

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443
  2    10.96.0.10:53     ClusterIP
  3    10.96.0.10:9153   ClusterIP

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS   BACKEND ADDRESS
  10.96.0.1:443     192.168.178.91:6443 (1)
                    0.0.0.0:0 (1) [ClusterIP, non-routable]
  10.96.0.10:53     0.0.0.0:0 (2) [ClusterIP, non-routable]
  10.96.0.10:9153   0.0.0.0:0 (3) [ClusterIP, non-routable]

  # ./cilium/cilium bpf lb list --frontends
  SERVICE ADDRESS   BACKEND ID
  10.96.0.10:9153   0 0 (3) [0x0 0x0]
  10.96.0.1:443     1 0 (1) [0x0 0x0]
                    0 1 (1) [0x0 0x0]
  10.96.0.10:53     0 0 (2) [0x0 0x0]

  # ./cilium/cilium bpf lb list --backends
  ID   BACKEND ADDRESS
  1    ANY://192.168.178.91:6443

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, api: Add natPolicy to service api for NAT46/64

This is mainly useful for service dumping / troubleshooting, but also
for restoration from the datapath.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bugtool: Add service --frontends, --backends, --revnat dump

Useful for debugging, so add to bugtool. Example:

  # ./cilium/cilium bpf lb list --frontends
  SERVICE ADDRESS   BACKEND ID
  10.96.0.10:53     0 (2) [flags: 0x0/0x0]
  10.96.0.1:443     0 (1) [flags: 0x0/0x0]
                    1 (1) [flags: 0x0/0x0]
  10.96.0.10:9153   0 (3) [flags: 0x0/0x0]

  # ./cilium/cilium bpf lb list --backends
  ID   BACKEND ADDRESS
  1    ANY://192.168.178.91:6443

  # ./cilium/cilium bpf lb list --revnat
  ID   BACKEND ADDRESS
  3    10.96.0.10:9153
  1    10.96.0.1:443
  2    10.96.0.10:53

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Dump v4/v6 backend and frontend maps

Add two new commands in order to be able to introspect map contents
for v4/v6 maps related to services and backends.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: Add initial NAT 6->4 datapath implementation

Manual 6->4 example setup with tcpdump extract of traffic being processed
on the LB node:

  # ./cilium/cilium service list
  ID   Frontend          Service Type   Backend
  1    10.96.0.1:443     ClusterIP      1 => 192.168.178.91:6443
  2    10.96.0.10:53     ClusterIP
  3    10.96.0.10:9153   ClusterIP

  # ./cilium/cilium service update --frontend="[2001:1620:665:0:1ac0:4dff:fe09:c164]:8080" --backends="1.1.1.1:80" --id 8 --k8s-external
  Creating new service with id '8'
  Added service with 1 backends

  # ./cilium/cilium service list
  ID   Frontend                                     Service Type   Backend
  1    10.96.0.1:443                                ClusterIP      1 => 192.168.178.91:6443
  2    10.96.0.10:53                                ClusterIP
  3    10.96.0.10:9153                              ClusterIP
  8    [2001:1620:665:0:1ac0:4dff:fe09:c164]:8080   ExternalIPs    1 => 1.1.1.1:80

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS                              BACKEND ADDRESS
  10.96.0.10:53                                0.0.0.0:0 (2) [ClusterIP, non-routable]
  [2001:1620:665:0:1ac0:4dff:fe09:c164]:8080   [::]:0 (8) [ExternalIPs, 46x64]
                                               [1.1.1.1]:80 (8)
  10.96.0.1:443                                192.168.178.91:6443 (1)
                                               0.0.0.0:0 (1) [ClusterIP, non-routable]
  10.96.0.10:9153                              0.0.0.0:0 (3) [ClusterIP, non-routable]

  # tcpdump -i enp5s0 port 8080 or port 80 -n
  [...]
  14:41:23.274051 IP6 2001:1620:665:0:1ac0:4dff:fe09:d5e6.43392 > 2001:1620:665:0:1ac0:4dff:fe09:c164.8080: Flags [S], seq 2445934322, win 43200, options [mss 1440,sackOK,TS val 3702225727 ecr 0,nop,wscale 9], length 0
  14:41:23.274074 IP 192.168.178.91.43392 > 1.1.1.1.80: Flags [S], seq 2445934322, win 43200, options [mss 1440,sackOK,TS val 3702225727 ecr 0,nop,wscale 9], length 0
  14:41:23.277411 IP 1.1.1.1.80 > 192.168.178.91.43392: Flags [S.], seq 3810710146, ack 2445934323, win 65535, options [mss 1436,nop,nop,sackOK,nop,wscale 10], length 0
  14:41:23.277422 IP6 2001:1620:665:0:1ac0:4dff:fe09:c164.8080 > 2001:1620:665:0:1ac0:4dff:fe09:d5e6.43392: Flags [S.], seq 3810710146, ack 2445934323, win 65535, options [mss 1436,nop,nop,sackOK,nop,wscale 10], length 0
  [...]

From external client, connecting to LB node which then performs the above IPv4
request to the backend:

  # curl --verbose "[2001:1620:665:0:1ac0:4dff:fe09:c164]:8080"
  *   Trying 2001:1620:665:0:1ac0:4dff:fe09:c164:8080...
  * TCP_NODELAY set
  * Connected to 2001:1620:665:0:1ac0:4dff:fe09:c164 (2001:1620:665:0:1ac0:4dff:fe09:c164) port 8080 (#0)
  > GET / HTTP/1.1
  > Host: [2001:1620:665:0:1ac0:4dff:fe09:c164]:8080
  > User-Agent: curl/7.68.0
  > Accept: */*
  >
  * Mark bundle as not supporting multiuse
  < HTTP/1.1 403 Forbidden
  < Date: Fri, 18 Feb 2022 14:41:32 GMT
  < Content-Type: text/plain; charset=UTF-8
  < Content-Length: 16
  < Connection: close
  < X-Frame-Options: SAMEORIGIN
  < Referrer-Policy: same-origin
  < Cache-Control: private, max-age=0, no-store, no-cache, must-revalidate, post-check=0, pre-check=0
  < Expires: Thu, 01 Jan 1970 00:00:01 GMT
  < Server: cloudflare
  < CF-RAY: 6df8044f4f1d01f4-ZRH
  <
  * Closing connection 0

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: For now skip E/W LB for NAT 4->6/6->4 services

For the time being, skip the NAT 4->6/6->4 for the E/W LB since it is out of
scope for this PR. Just pass the traffic through instead.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Support restoration for NAT 4->6 upon Maglev LUT recreation

When switching from `random` to `maglev` or when the maglev LUT size changes
between lb-only agent runs and services are being restored, we also need to
handle NAT 4->6 cases, otherwise they are not propagated to the v6 outer maglev
map, and then connectivity will not work.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: Fix ipv6_to_ipv4 csum for carry in ICMPv6

Lets add a proper csum_sub(), and use it for the icmp6 to icmp4 conversion.
Plain sub does not take carry into account, and can thus lead to broken
csums as reported by Rumen.

Reported-by: Rumen Telbizov (via Slack)
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Dump Maglev LUT with information on revnat id origin

Dump info whether the key is from v4 or v6 outer maglev map. This is very
useful debug info to see if a key is present in both outer tables. Before
this change it is only possible to check via bpftool given the go hashtab
we build has the same key otherwise.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, tests: Add NAT46 and NAT64 upsert test cases

Add various test cases for the agent's service management as well as the
BPF service map for NAT46/64 entries.

  root@zh-lab-node-1:~/go/src/github.com/cilium/cilium/pkg/service# go test -v -check.vv -check.f TestUpsertAndDeleteServiceNat
  === RUN   Test
  START: service_test.go:103: ManagerTestSuite.TestUpsertAndDeleteServiceNat46
  START: service_test.go:38: ManagerTestSuite.SetUpTest
  PASS: service_test.go:38: ManagerTestSuite.SetUpTest	0.000s

  START: service_test.go:62: ManagerTestSuite.TearDownTest
  PASS: service_test.go:62: ManagerTestSuite.TearDownTest	0.000s

  PASS: service_test.go:103: ManagerTestSuite.TestUpsertAndDeleteServiceNat46	0.000s

  START: service_test.go:109: ManagerTestSuite.TestUpsertAndDeleteServiceNat64
  START: service_test.go:38: ManagerTestSuite.SetUpTest
  PASS: service_test.go:38: ManagerTestSuite.SetUpTest	0.000s

  START: service_test.go:62: ManagerTestSuite.TearDownTest
  PASS: service_test.go:62: ManagerTestSuite.TearDownTest	0.000s

  PASS: service_test.go:109: ManagerTestSuite.TestUpsertAndDeleteServiceNat64	0.000s

  OK: 2 passed
  --- PASS: Test (0.00s)
  PASS
  ok  	github.com/cilium/cilium/pkg/service	0.009s
  [...]

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: Move NodePort specific fallback defines to stub.h

Just move them to a dummy header file to it's less noise in nodeport.h
and add a fallback for IPV4_DIRECT_ROUTING when v4 is not enabled.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: Add ctx_change_proto implementation for XDP context

Add an implementation for the XDP context, so NAT46/64 can be used together
with the standalone XDP L4LB.

Tested that NAT46/64 works on standalone LB with:

  # ./daemon/cilium-agent --enable-ipv4=true --enable-ipv6=true
      --disable-envoy-version-check=true --tunnel=disabled
      --kube-proxy-replacement=strict --enable-ipv4-masquerade=false
      --enable-ipv6-masquerade=false --enable-hubble=true
      --enable-recorder=true --datapath-mode=lb-only
      --bpf-lb-algorithm=maglev --bpf-lb-maglev-table-size=2039
      --bpf-lb-acceleration=native

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, api: Extend status API with state for NAT46/64

Extend the Cilium status API with a dump for NAT46/64 which is mainly useful
for service dumping / troubleshooting.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, status: Wire up NAT46/64 status dump

Wire the NAT46/64 status dump up to the daemon / client packages.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, tests: Add NAT46x64 standalone LB runtime test suite

Add a GH test suite for the standalone L4LB to exercise the NAT46 and NAT64
BPF code at runtime. This also adds a PCAP recorder complexity test when the
NAT code is enabled under XDP, and 32 rules for each IP family are loaded.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb commented on api/v1/models/service_spec.go:
```diff
@@ -152,6 +152,10 @@ type ServiceSpecFlags struct {
 	// Service namespace  (e.g. Kubernetes namespace)
 	Namespace string `json:"namespace,omitempty"`
 
+	// Service protocol NAT policy
```
Can't comment the commit msg, but how this API sec field is going to be used for:
> but also for restoration from the datapath.

borkmann replied:
The API field is used in the dump. Iv'e added it to the json output given it's more verbose and this way it's easier to grasp. The regular service list is intuitive from the addresses, imho, but we could also add it there.
```
# ./cilium/cilium service list -o=json
[
  {
    "spec": {
      "backend-addresses": [
        {
          "ip": "2a03:2880:f11c:8183:face:b00c:0:25de",
          "port": 80
        }
      ],
      "flags": {
        "natPolicy": "Nat46",
        "trafficPolicy": "Cluster",
        "type": "ExternalIPs"
      },
      "frontend-address": {
        "ip": "192.168.178.91",
        "port": 8082,
        "scope": "external"
      },
      "id": 11
    },
    "status": {
      "realized": {
        "backend-addresses": [
          {
            "ip": "2a03:2880:f11c:8183:face:b00c:0:25de",
            "port": 80
          }
        ],
        "flags": {
          "natPolicy": "Nat46",
          "trafficPolicy": "Cluster",
          "type": "ExternalIPs"
        },
        "frontend-address": {
          "ip": "192.168.178.91",
          "port": 8082,
          "scope": "external"
        },
        "id": 11
      }
    }
  },
  {
    "spec": {
      "backend-addresses": [
        {
          "ip": "1.1.1.1",
          "port": 80
        }
      ],
      "flags": {
        "natPolicy": "Nat64",
        "trafficPolicy": "Cluster",
        "type": "ExternalIPs"
      },
      "frontend-address": {
        "ip": "2001:1620:665:0:1ac0:4dff:fe09:c164",
        "port": 8080,
        "scope": "external"
      },
      "id": 8
    },
    "status": {
      "realized": {
        "backend-addresses": [
          {
            "ip": "1.1.1.1",
            "port": 80
          }
        ],
        "flags": {
          "natPolicy": "Nat64",
          "trafficPolicy": "Cluster",
          "type": "ExternalIPs"
        },
        "frontend-address": {
          "ip": "2001:1620:665:0:1ac0:4dff:fe09:c164",
          "port": 8080,
          "scope": "external"
        },
        "id": 8
      }
    }
  }
]
```

borkmann replied:
The restoration holds the NAT46x64 bit in the flags2 field in the svc map, and reconstructs svc from there.

brb commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -67,22 +68,17 @@ type UpsertServiceParams struct {
 
-	if p.ID == 0 {
-		return fmt.Errorf("Invalid svc ID 0")
+	backendsOk := ipv6 || !ipv6 && p.NatPolicy != loadbalancer.SVCNatPolicyNat46
```
This requires a code comment.

borkmann replied:
Agree, added comment about storage of backends.

brb commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -67,22 +68,17 @@ type UpsertServiceParams struct {
-// backend entries that were added to the BPF map so that the function can remove
-// obsolete ones.
-func (lbmap *LBBPFMap) UpsertService(p *UpsertServiceParams) error {
+func (lbmap *LBBPFMap) upsertServiceProto(p *UpsertServiceParams, ipv6 bool) error {
```
It would be useful if you could add a comment how nat46x64 entries are being stored.

borkmann replied:
I've added an extensive explanation incl various dumps to the commit msg ("cilium: Initial implementation of agent NAT46/64 handling"). (https://github.com/cilium/cilium/pull/18779/commits/d70a8c1e772ce02838ac6d520e16fe193241f504)

brb commented on pkg/service/service.go:
```diff
@@ -37,9 +37,9 @@ type LBMap interface {
 	UpsertMaglevLookupTable(uint16, map[string]lb.BackendID, bool) error
 	IsMaglevLookupTableRecreated(bool) bool
-	DeleteService(lb.L3n4AddrID, int, bool) error
+	DeleteService(lb.L3n4AddrID, int, bool, lb.SVCNatPolicy) error
```
Why do we need to pass the NAT pol? We could store it in the svc handler's state.

brb commented on pkg/service/service.go:
```diff
@@ -339,16 +341,39 @@ func (s *Service) UpsertService(params *lb.SVC) (bool, lb.ID, error) {
 
-	ipv6Svc := params.Frontend.IsIPv6()
-	if ipv6Svc && !option.Config.EnableIPv6 {
+	// Backends must either be the same IP proto as the frontend, or can be of
```
Shall we check for Option.EnableNAT46x64 too?

borkmann replied:
agree, done.

brb commented on pkg/service/service.go:
```diff
@@ -781,17 +807,28 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, onlyLocalBackends bool,
 	}
 
 	// Upsert service entries into BPF maps
+	natPolicy := lb.SVCNatPolicyNone
```
We should be able to derive the NAT pol from `svc.svcNatPolicy`, shouldn't we?

brb commented on pkg/service/service.go:
```diff
@@ -804,15 +841,27 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, onlyLocalBackends bool,
 	}
+	if natPolicy == lb.SVCNatPolicyNat64 {
+		// Backends have been added to the v4 backend map, but we now also need
+		// to add them to the v6 backend map as v4-in-v6 address.
```
Please give more details for why.

borkmann replied:
I've expanded the comment with more details on why it's needed.

brb commented on pkg/service/service.go:
```diff
@@ -1005,10 +1050,7 @@ func (s *Service) deleteServiceLocked(svc *svcInfo) error {
-		if err := s.lbmap.DeleteBackendByID(id, ipv6); err != nil {
-			return err
-		}
+		s.lbmap.DeleteBackendByID(id)
```
Why ignoring errors?

borkmann replied:
The id is unique across both v4/v6 backends. Only based on the ID we don't know whether a backend is in the v4 or v6 map, so we need to attempt to delete it from both. Also, when we remove orphan backends in general I don't think we should bail from that with error.

brb replied:
If we store the NAT policy in `svcInfo`, then we could be able to determine which backends need exactly to be deleted.

If someone extends `DeleteBackendByID()`, then the err ignore might get into our way. Usually when I see a function returning an error, my expectation that the error return is going to be handled by callers.

borkmann replied:
For the case of `deleteOrphanBackends` right now it's actually worse. If for some reason the backend map returns `ENOENT` if the to-be-deleted backend doesn't exist anymore in the BPF map (non-fatal error, and the only one BPF map can return), then we cancel all remaining deletion (of unrelated backends) and leaving stale backends with refcount 0 intact. Caller won't handle anything, just dumping a warn log message.

brb commented on pkg/service/service_test.go:
```diff
@@ -100,6 +100,98 @@ func (m *ManagerTestSuite) TestUpsertAndDeleteServiceWithoutIPv6(c *C) {
 	m.testUpsertAndDeleteService(c)
 }
 
+func (m *ManagerTestSuite) TestUpsertAndDeleteServiceNat46(c *C) {
```
We need a test for the NAT pol restoration.

borkmann replied:
The last test suite covers restoration for both NAT46 and NAT64, by comparing service dump between restarts and doing runtime tests, also for Maglev vs Random selection algs.

brb replied:
I can't find, which one it is? I was refering to `TestRestoreServices()`.

brb replied:
Also, it would be good to have a coverage for changing svc types (none <=> nat46 <=> nat64).

brb commented on bpf/lib/lb.h:
```diff
@@ -1301,6 +1323,17 @@ lb4_update_affinity_by_netns(const struct lb4_service *svc __maybe_unused,
 }
 
+static __always_inline int
+lb4_to_lb6(struct __ctx_buff *ctx, const struct iphdr *ip4, int l3_off)
```
To me the function name reads as it's doing "LB from v4 to v6", while it changes the v4 request to v6 w/o doing any loadbalancing.

brb commented on bpf/lib/lb.h:
```diff
@@ -1059,6 +1069,22 @@ bool lb4_src_range_ok(const struct lb4_service *svc __maybe_unused,
 
+static __always_inline void
+lb4_populate_ports(struct __ctx_buff *ctx, struct ipv4_ct_tuple *tuple, int off)
+{
```
To prevent an accidental usage with ICMP, maybe we could add the assertion for != IPPROTO_ICMP?

borkmann replied:
Great point, I've changed it into:
```                if (!lb4_populate_ports(ctx, &tuple, l4_off) &&
                    snat_v6_has_v4_match(&tuple)) {
                        ret = lb4_to_lb6(ctx, ip4, l3_off);
                        if (ret)
                                return ret;
                        ep_tail_call(ctx, CILIUM_CALL_IPV6_NODEPORT_NAT);
                } else {
```
... and return 0 if we successfully populated ports for udp/tcp, everything else will error and take the regular CILIUM_CALL_IPV4_NODEPORT_NAT path.

brb commented on bpf/lib/lb.h:
```diff
@@ -1059,6 +1069,22 @@ bool lb4_src_range_ok(const struct lb4_service *svc __maybe_unused,
 }
 
+static __always_inline void
+lb4_populate_ports(struct __ctx_buff *ctx, struct ipv4_ct_tuple *tuple, int off)
```
Shouldn't we move this function to lib/conntrack.h?

borkmann replied:
(We don't access its maps and only relevant to LB, so I thought to keep it here.)

brb commented on bpf/lib/nodeport.h:
```diff
@@ -840,6 +849,9 @@ static __always_inline int nodeport_lb6(struct __ctx_buff *ctx,
 		} else {
 			ctx_store_meta(ctx, CB_NAT, NAT_DIR_EGRESS);
+			ctx_store_meta(ctx, CB_NAT_46X64,
+				       !is_v4_in_v6(&key.address) &&
```
Maybe a small comment saying that the same code path is hit for NAT46, thus `!is_v4_in_v6()`.

borkmann replied:
Agree, added comment with explanation.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -674,6 +677,11 @@ int tail_nodeport_nat_ipv6(struct __ctx_buff *ctx)
+			goto drop_err;
+	}
 
 	ret = fib_lookup(ctx, &fib_params.l, sizeof(fib_params), 0);
```
Are we still OK to do the FIB lookup for the v6 addrs?

borkmann replied:
Fixed, and turned out that locally this wasn't an issue, but in the CI runner, the fib lookup was failing due to that. So yeah, v4 is now done in the updated code.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1777,10 +1789,17 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
 		ctx_store_meta(ctx, CB_NAT, NAT_DIR_INGRESS);
 		ctx_store_meta(ctx, CB_SRC_IDENTITY, src_identity);
-		ep_tail_call(ctx, CILIUM_CALL_IPV4_NODEPORT_NAT);
+		lb4_populate_ports(ctx, &tuple, l4_off);
```
I am bit missing how in the case of NAT64 we can enter this code path. In any case, a comment would be useful.

borkmann replied:
NAT64 basically transforms a v6 req into a v4 req, so the LB to backend path is v4, and then this will be his via reply from backend. Will add a comment.

brb commented on .github/workflows/tests-nat46x64.yaml:
```diff
@@ -0,0 +1,233 @@
+name: Cilium L4LB NAT-46x64 (ci-nat46x64)
```
Have you looked how to avoid the boilerplate with test-l4lb.yaml? Maybe we could add to the test-l4lb an additional step to run the tests? I doubt that the k8s Service native support for NAT46x64 will appear soon (does k8s even allow mixing ip families in Service?), so we will need to live with the tests for awhile.

borkmann replied:
Given the completely different setup, yaml, agent and routing configs, I went for new one which felt cleaner.

brb replied:
But if the `test/l4lb/test.sh` does the proper cleanup, then the only change is required is to extend the `.github/workflows/test-l4lb.yaml`:
```
- name: Run tests
        run: |
          ssh default "sudo /bin/sh -c 'cd /vagrant/test/l4lb && ./test.sh ${{ github.repository_owner}} ${{ steps.vars.outputs.sha }}'"
          ssh default "sudo /bin/sh -c 'cd /vagrant/test/nat46x64 && ./test.sh ${{ github.repository_owner}} ${{ steps.vars.outputs.sha }}'"
```
I could take a look into making the former to properly cleanup the VM.

brb reviewed (changes requested):
Nice work. Did the first round of reviews (will do the 2nd after the requested code comments appear).

I think this PR would benefit from a cover letter which would explain the following:
- How the NAT46x64 services are provisioned. What is the content of lbmaps for each NAT policy.
- Packet path for NAT46.
- Packet path for NAT64.

This would help the reviewers (incl. myself) to more quickly understand the changes.

borkmann committed:
bugtool: Add verbose service json dump

Useful to gather all API data from the agent PoV, e.g.:

  # ./cilium/cilium service list -o=json
  [
    {
      "spec": {
        "backend-addresses": [
          {
            "ip": "2a03:2880:f11c:8183:face:b00c:0:25de",
            "port": 80
          }
        ],
        "flags": {
          "natPolicy": "Nat46",
          "trafficPolicy": "Cluster",
          "type": "ExternalIPs"
        },
        "frontend-address": {
          "ip": "192.168.178.91",
          "port": 8082,
          "scope": "external"
        },
        "id": 11
      },
      "status": {
        "realized": {
          "backend-addresses": [
            {
              "ip": "2a03:2880:f11c:8183:face:b00c:0:25de",
              "port": 80
            }
          ],
          "flags": {
            "natPolicy": "Nat46",
            "trafficPolicy": "Cluster",
            "type": "ExternalIPs"
          },
          "frontend-address": {
            "ip": "192.168.178.91",
            "port": 8082,
            "scope": "external"
          },
          "id": 11
        }
      }
    },
    {
      "spec": {
        "backend-addresses": [
          {
            "ip": "1.1.1.1",
            "port": 80
          }
        ],
        "flags": {
          "natPolicy": "Nat64",
          "trafficPolicy": "Cluster",
          "type": "ExternalIPs"
        },
        "frontend-address": {
          "ip": "2001:1620:665:0:1ac0:4dff:fe09:c164",
          "port": 8080,
          "scope": "external"
        },
        "id": 8
      },
      "status": {
        "realized": {
          "backend-addresses": [
            {
              "ip": "1.1.1.1",
              "port": 80
            }
          ],
          "flags": {
            "natPolicy": "Nat64",
            "trafficPolicy": "Cluster",
            "type": "ExternalIPs"
          },
          "frontend-address": {
            "ip": "2001:1620:665:0:1ac0:4dff:fe09:c164",
            "port": 8080,
            "scope": "external"
          },
          "id": 8
        }
      }
    }
  ]

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb commented on pkg/service/service.go:
```diff
@@ -804,15 +844,31 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, onlyLocalBackends bool,
 			activeBackendsCount++
 		}
 	}
+	if natPolicy == lb.SVCNatPolicyNat64 {
```
We have the following actors in the svc provisioning pipeline:
- `pkg/service` - responsible for API and k8s svc provisioning requests
- `pkg/maps/lbmap` - responsible for updating the relevant BPF maps.

Ideally, only the latter should know about the low level details of the service representation in the datapath. IMHO, adding to the v4 and v6 backend maps in the case of NAT64 is an example of such a low level details. So maybe we could move the provisioning of both to the `lbmap`?

borkmann committed:
bpf: Only enable NAT46/64 for 5.4+ kernels

For 4.19 kernels, NAT46/64 code is generating more than 4096 insns inside
the tail_nodeport_nat_ipv6() tail called program:

  [...]
  libbpf: load bpf program failed: Argument list too long
  libbpf: Program too large (4109 insns), at most 4096 insns
  libbpf: failed to load program 'tail_nodeport_nat_ipv6'
  [...]

Thus, restrict NAT46/64 enablement only for kernels with large insn limit.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb reviewed (approved):
:rocket:

julianwiedmann commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -130,26 +132,51 @@ func (lbmap *LBBPFMap) UpsertService(p *UpsertServiceParams) error {
+	}
+	if p.NatPolicy == loadbalancer.SVCNatPolicyNat46 {
+		if err := lbmap.upsertServiceProto(p, false); err != nil {
+			return err
```
hi @borkmann - rather late to the party, but need one silly clarification. For a NAT46 service, why do we create an additional Maglev table here for the IPv4 dummy frontend? I thought the backend selection for NAT46 happens in the IPv6 codepath...?

brb replied:
@julianwiedmann That's a good question. I will defer to @borkmann for an answer, but to me it seems a bug. Also, that should explain the failures in the L4LB GH action https://github.com/cilium/cilium/pull/20625.

borkmann replied:
We shouldn't, the IPv4 to IPv6 has only dummy IPv4 service entry (0 backends) given it will recircle the packet into the IPv6 path. @julianwiedmann do we create a full-blown table?

borkmann replied:
This one seems to be missing:
```
diff --git a/pkg/maps/lbmap/lbmap.go b/pkg/maps/lbmap/lbmap.go
index db6b83521e..6b1f68c402 100644
--- a/pkg/maps/lbmap/lbmap.go
+++ b/pkg/maps/lbmap/lbmap.go
@@ -87,15 +87,14 @@ func (lbmap *LBBPFMap) upsertServiceProto(p *datapathTypes.UpsertServiceParams,
                backends = p.PreferredBackends
        }
 
-       if p.UseMaglev && len(backends) != 0 {
-               if err := lbmap.UpsertMaglevLookupTable(p.ID, backends, ipv6); err != nil {
-                       return err
+       if backendsOk {
+               if p.UseMaglev && len(backends) != 0 {
+                       if err := lbmap.UpsertMaglevLookupTable(p.ID, backends, ipv6); err != nil {
+                               return err
+                       }
                }
-       }
 
-       if backendsOk {
                backendIDs := p.GetOrderedBackends()
-
                for _, backendID := range backendIDs {
                        if backendID == 0 {
                                return fmt.Errorf("Invalid backend ID 0")
```

borkmann replied:
(Pushed here https://github.com/cilium/cilium/pull/20648)
